### ISSUE TITLE
Why time cost for ScaledDotProduct with sparse mask is not fast as expected?

### ISSUE THREAD

**Issue by zihanlalala** [facebookresearch/xformers#441]
# ❓ Questions and Help
I have tested the time cost of ScaledDotProduct with masks having different sparsity and compare them with simple original attention. I find that even mask with 70% sparsity is slower than original attention, although theoretical only 30% computation is calculated.
Does my testing code wrong? 
Here is my test code, which refers to HOWTO.md in the project:

```
import torch
try:
    from xformers.components.attention import build_attention
    from xformers.components.attention._sputnik_sparse import SparseCS
    _xformers_available = True
except ImportError:
    _xformers_available = False

assert _xformers_available

bsz = 64
seq_len = 2048
seq_len = 2048
emb_dim = 512

config = {
    "name": "scaled_dot_product"
}

xformer_attention = build_attention(config).to("cuda")

def rand_data(sparsity):
    q = torch.rand(size=[bsz, seq_len, emb_dim], dtype=torch.float32, device="cuda")
    ## ones are computed. 
    mask = torch.rand(size=[seq_len, seq_len], dtype=torch.float32, device="cuda") > sparsity
    return q, mask

#########################################
# Bonus: compare the memory use vs dense:
def mem_use(fn, title, q, mask, sparsity):
    # bookeeping
    import time
    torch.cuda.empty_cache()
    torch.cuda.reset_peak_memory_stats()

    start = time.time()
    # actually run the function, run 10 times to avoid noise
    for _ in range(10):
        q = fn(q=q, k=q, v=q, att_mask=mask)
    torch.cuda.synchronize()
    stop = time.time()

    # now report
    max_memory = torch.cuda.max_memory_allocated() // 2 ** 20
    print(f"{title} - sparsity is {sparsity} - Peak memory use: {max_memory}MB - {round((stop-start)*1e6)/1e3}ms")
    return round((stop-start)*1e6)/1e3

### simple attention using torch.bmm
def simple_attention(q,k,v,att_mask):
    attn_weights = torch.bmm(q, k.transpose(1,2))
    attn_weights = torch.softmax(attn_weights, dim=-1)
    attn_weights = torch.masked_fill(attn_weights, att_mask, float(-1e8))
    output = torch.bmm(attn_weights, v)
    assert list(output.size()) == [bsz, seq_len, emb_dim]
    return output

## repeat multiple times 
for _ in range(3):
    ## different sparsity time cost
    for sparsity in range(10,0,-1):
        q, mask = rand_data(sparsity=sparsity/10)
        mask = SparseCS(mask, device="cuda")
        t = mem_use(xformer_attention, "xformer_attention", q=q, mask=mask, sparsity=sparsity/10)
    ## simple attention
    q, mask = rand_data(sparsity=0)
    t = mem_use(simple_attention, "simple_attention", q=q, mask=mask, sparsity=0)
    print("\n)

```

I repeat it three times, and here is the output:
```
xformer_attention - sparsity is 1.0 - Peak memory use: 768MB - 9.043ms
xformer_attention - sparsity is 0.9 - Peak memory use: 979MB - 164.253ms
xformer_attention - sparsity is 0.8 - Peak memory use: 1191MB - 336.672ms
xformer_attention - sparsity is 0.7 - Peak memory use: 1404MB - 368.312ms
xformer_attention - sparsity is 0.6 - Peak memory use: 1615MB - 464.942ms
xformer_attention - sparsity is 0.5 - Peak memory use: 1826MB - 555.514ms
xformer_attention - sparsity is 0.4 - Peak memory use: 2038MB - 638.45ms
xformer_attention - sparsity is 0.3 - Peak memory use: 2249MB - 728.075ms
xformer_attention - sparsity is 0.2 - Peak memory use: 2462MB - 810.209ms
xformer_attention - sparsity is 0.1 - Peak memory use: 2673MB - 886.492ms
simple_attention - sparsity is 0 - Peak memory use: 2564MB - 1941.696ms

xformer_attention - sparsity is 1.0 - Peak memory use: 768MB - 7.797ms
xformer_attention - sparsity is 0.9 - Peak memory use: 979MB - 166.427ms
xformer_attention - sparsity is 0.8 - Peak memory use: 1191MB - 281.606ms
xformer_attention - sparsity is 0.7 - Peak memory use: 1403MB - 411.525ms
xformer_attention - sparsity is 0.6 - Peak memory use: 1615MB - 462.009ms
xformer_attention - sparsity is 0.5 - Peak memory use: 1827MB - 554.949ms
xformer_attention - sparsity is 0.4 - Peak memory use: 2037MB - 635.765ms
xformer_attention - sparsity is 0.3 - Peak memory use: 2248MB - 725.433ms
xformer_attention - sparsity is 0.2 - Peak memory use: 2462MB - 810.413ms
xformer_attention - sparsity is 0.1 - Peak memory use: 2673MB - 886.657ms
simple_attention - sparsity is 0 - Peak memory use: 2564MB - 336.262ms

xformer_attention - sparsity is 1.0 - Peak memory use: 768MB - 13.247ms
xformer_attention - sparsity is 0.9 - Peak memory use: 980MB - 164.073ms
xformer_attention - sparsity is 0.8 - Peak memory use: 1191MB - 267.765ms
xformer_attention - sparsity is 0.7 - Peak memory use: 1402MB - 368.902ms
xformer_attention - sparsity is 0.6 - Peak memory use: 1615MB - 467.156ms
xformer_attention - sparsity is 0.5 - Peak memory use: 1826MB - 555.288ms
xformer_attention - sparsity is 0.4 - Peak memory use: 2037MB - 721.594ms
xformer_attention - sparsity is 0.3 - Peak memory use: 2248MB - 724.735ms
xformer_attention - sparsity is 0.2 - Peak memory use: 2463MB - 805.58ms
xformer_attention - sparsity is 0.1 - Peak memory use: 2673MB - 886.753ms
simple_attention - sparsity is 0 - Peak memory use: 2564MB - 334.988ms
```

I find that at the first time simple attention cost much but later the cost is even smaller than xformer_attention with mask sparsity 0.7. I wonder why simple attention cost is much higher at the first time and does my test code wrong?
I test it on A100 with one gpu,  Driver Version: 450.142.00   CUDA Version: 11.3.

By the way, sparse mask should be converted manually to **SparseCS** to enable Sparse ScaledDotProduct, while the HOWTO.md says sparse ScaledDotProduct will be enabled as soon as the mask is sparse enough.

By the way, I find that in file sddmm.cu, the block dim is relatively small,
```
  dim3 grid_dim(
      std::ceil(static_cast<float>(m) / kBlockItemsY),
      std::ceil(static_cast<float>(n) / kBlockItemsX),
      batch_size);
  dim3 block_dim(kBlockWidth, kBlockItemsY, 1);
```
where kBlockWidth * kBlockItemsY is 32, which means there are only 32 threads per block, while in cuda programming guide, 256 threads per block is a common use. Does this affect the performance?

**Comment by fmassa** [comment]
Hi,

There are four points to your question, which I'm addressing separately

## Unstable benchmark numbers

The reason why your first benchmarking for the vanilla attention is very large is because you don't perform any warmup in your benchmark before measuring the time, so you are also measuring the time for the first allocation of memory for the large attention matrix.

Adding a couple of iterations in the beginning of your `mem_use` function to launch a few iterations will make things stable.

That being said, I would encourage you to use `torch.utils.benchmark`, which performs all of this to you, including CUDA synchronization and can also do comparison between runs. This is what we use in our benchmarks in xformers as well, see `xformers/benchmarks/` for some examples. We btw benchmark the sparse operators that are being used in https://github.com/facebookresearch/xformers/blob/main/xformers/benchmarks/benchmark_core.py

Also, a couple of other thoughts:
- try not re-using the same `q` as both input and output of the function, as this could affect numbers if for some reason there are `nan`s that appear
- You forgot to divide by `sqrt(dim)`, which is a cheap operation but that should be taken into account if you would like to compare apples with apples

## Dense attention faster than sparse attention

It is expected that at ~70% sparsity your code will be roughly as fast as with dense attention. The reason is that loading the indices of the sparse locations has a cost, which is only amortized at reasonably high sparsity values. That being said, 80% sparsity is not that much for sparse operation standards, and can indeed bring faster runtimes than vanilla dense attention (without using TensorCores).

## Thread-block sizes

There is a trade-off between the number of used registers in a thread and the ideal number of threads to launch per block.
For this case in particular, I just kept the original hyperparameters that were setup by sputnik kernels and haven't done ablations on this (I've only extended their kernels to support batches natively).

## Outdated documentation

> By the way, sparse mask should be converted manually to SparseCS to enable Sparse ScaledDotProduct, while the HOWTO.md says sparse ScaledDotProduct will be enabled as soon as the mask is sparse enough.

Yes, this is a bug that should be fixed in the HOWTO. Can you send a PR fixing it?


I believe I've answered your questions and as such I'm closing this issue, but let us know if you have further questions.
